tests/TestCase/Database/Log/QueryLoggerTest.php

```diff
@@ -128,6 +128,26 @@ public function testStringInterpolationNamed()
     }
 
     /**
+     * Tests that placeholders are replaced with correctly escaped strings
+     *
+     * @return void
+     */
+    public function testStringInterpolationSpecialChars()
+    {
+        $logger = $this->getMockBuilder('\Cake\Database\Log\QueryLogger')
+            ->setMethods(['_log'])
+            ->getMock();
+        $query = new LoggedQuery;
+        $query->query = 'SELECT a FROM b where a = :p1 AND b = :p2 AND c = :p3 AND d = :p4';
+        $query->params = ['p1' => '$2y$10$dUAIj', 'p2' => '$0.23', 'p3' => 'a\\0b\\1c\\d', 'p4' => "a'b"];
+
+        $logger->expects($this->once())->method('_log')->with($query);
+        $logger->log($query);
+        $expected = "duration=0 rows=0 SELECT a FROM b where a = '\$2y\$10\$dUAIj' AND b = '\$0.23' AND c = 'a\\\\0b\\\\1c\\\\d' AND d = 'a''b'";
```

Comment:
The final SQL is 
`SELECT a FROM b where a = '$2y$10$dUAIj' AND b = '$0.23' AND c = 'a\\0b\\1c\\d' AND d = 'a''b'`

And the final strings after SQL unescape are `c = a\0b\1c\d` and `d = a'b` - c has backslash and d has single-quote and matches `'p3' => 'a\\0b\\1c\\d', 'p4' => "a'b"` after PHP unescape.

If it would be `\\` instead of `\\\\`, SQL would get `\0` which would result to null-byte after SQL unescape.

This is light version of `PDO::quote()`  - only escapes `\` to `\\` and `'` to `''`.
While it is not complete solution for safe SQL escape, but developer needs less modifications before testing SQL elsewhere.

The `'` escape to `''` is supported by all 4 CakePHP core SQL drivers.

The `\` escape to `\\` is supported by MySQL and PostgreSQL, but in Sqlite and SQL Server the `\` in string literals has no special meaning.

---

1. The commit fixes preg_replace part

```
$replacements = [
    '$' => '\\$',
    '\\' => '\\\\',
];
```

2. Then enhances generated SQL string quoting

```
$replacements = [
    '\\\\' => '\\\\\\\\',
    "'" => "''",
];
```

Should I:
- leave only part 1
- leave part 1, and then escape only `'` to `''` (MySQL/PostgreSQL developer will need escape `\` manually)
- or stay with current commit - leave part 1, and then escape both `'` and `\` (Sqlite/SQL Server developer will need revert `\\` back to `\` manually)?